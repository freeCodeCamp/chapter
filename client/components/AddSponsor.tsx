import React, { useState } from 'react';
import { useForm, Controller } from 'react-hook-form';
import {
  Button,
  TextField,
  makeStyles,
  Select,
  MenuItem,
  InputLabel,
  FormControl,
} from '@material-ui/core';

const useStyles = makeStyles(() => ({
  form: {
    display: 'flex',
    flexDirection: 'column',
    maxWidth: '25%',
  },
  item: {
    marginTop: '20px',
  },
  responseDiv: {
    margin: '15px 0',
  },
}));

const AddSponsor: React.FC = () => {
  const [responseMsg, setResponseMsg] = useState('');
  const styles = useStyles();

  const { control, handleSubmit } = useForm();

  // TODO: Get data from store
  // const eventId = useSelector(state => state.selectedChapter.eventId);
  // const chapterId = useSelector(state => state.selectedChapter.id);

  const onSubmit = async data => {
    try {
      // await dispatch(sponsorActions.submit(eventId, chapterId));
      setResponseMsg(`${data.name} has been added as a ${data.type} sponsor.`);
    } catch (e) {
      setResponseMsg('Uh oh, something went wrong.');
      // TODO: more descriptive error messages
    }
  };

  return (
    <>
      {responseMsg && <div className={styles.responseDiv}>{responseMsg}</div>}
      <form onSubmit={handleSubmit(onSubmit)} className={styles.form}>
        <FormControl className={styles.item}>
          <Controller
            control={control}
            as={
              <TextField
                name="name"
                type="text"
                label="Sponsor Name"
                placeholder="Glitter and Sparkle Co"
              />
            }
            name="name"
            options={{ required: true }}
          />
        </FormControl>
        <FormControl className={styles.item}>
          <Controller
            control={control}
            as={
              <TextField
                label="Sponsor Website"
                name="website"
                type="text"
                placeholder="www.glitter.co"
              />
            }
            name="website"
            options={{ required: true }}
          />
        </FormControl>
        <FormControl className={styles.item}>
          <InputLabel id="sponsor-type-label">Sponsor Type</InputLabel>
          <Controller
            control={control}
            as={
              <Select labelId="sponsor-type-label">
                <MenuItem value={'FOOD'}>Food</MenuItem>
                <MenuItem value={'BEVERAGE'}>Beverage</MenuItem>
                <MenuItem value={'OTHER'}>Other</MenuItem>
              </Select>
            }
            name="type"
            defaultValue={'OTHER'}
            options={{ required: true }}
          />
        </FormControl>
        <Button
          className={styles.item}
          variant="contained"
          color="primary"
          type="submit"
        >
          Add Sponsor
        </Button>
      </form>
    </>
  );
};

export default AddSponsor;
