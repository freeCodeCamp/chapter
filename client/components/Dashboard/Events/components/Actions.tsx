import React from 'react';
import { Button } from '@material-ui/core';
import Link from 'next/link';

import { eventActions } from 'client/store/actions';
import useConfirm from 'client/hooks/useConfirm';
import useThunkDispatch from 'client/hooks/useThunkDispatch';
import { IEventModal } from 'client/store/types/events';

interface IActionsProps {
  event: IEventModal;
  onDelete?: () => void;
}

const Actions: React.FC<IActionsProps> = ({ event, onDelete }) => {
  const dispatch = useThunkDispatch();

  const [confirmCancel, clickCancel] = useConfirm(() =>
    dispatch(eventActions.cancelEvent(1, event.id)),
  );
  const [confirmRemove, clickRemove] = useConfirm(() => {
    dispatch(eventActions.removeEvent(1, event.id));
    onDelete && onDelete();
  });

  return (
    <>
      {!event.canceled && (
        <Button onClick={clickCancel}>
          {confirmCancel ? 'Are you sure?' : 'Cancel'}
        </Button>
      )}
      <Button onClick={clickRemove}>
        {confirmRemove ? 'Are you sure?' : 'delete'}
      </Button>
      <Link
        href={`/dashboard/events/[id]/edit`}
        as={`/dashboard/events/${event.id}/edit`}
      >
        <a style={{ marginLeft: '10px' }}>Edit</a>
      </Link>
    </>
  );
};

export default Actions;
