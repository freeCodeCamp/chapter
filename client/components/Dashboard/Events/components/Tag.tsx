import React from 'react';
import { makeStyles } from '@material-ui/core';

import { ITagModal } from 'client/store/types/events';

const useStyles = makeStyles({
  tag: {
    borderRadius: '8px',
    padding: '4px 12px',
    backgroundColor: '#22c98c',
    margin: '8px 2px',
    color: '#fff',
    fontWeight: 600,
  },
  tagsContainer: { display: 'flex' },
});

const Tags: React.FC<{ tags?: ITagModal[] }> = ({ tags }) => {
  const style = useStyles();

  return tags ? (
    <div className={style.tagsContainer}>
      {tags.map(tag => (
        <div className={style.tag} key={`tag-${tag.id}`}>
          {tag.name}
        </div>
      ))}
    </div>
  ) : null;
};

export default Tags;
