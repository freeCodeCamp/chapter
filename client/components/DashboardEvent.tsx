import React from 'react';
import { Card, Typography } from '@material-ui/core';

import ProgressCardContent from './ProgressCardContent';
import { IEventModal } from 'client/store/types/events';
import Link from 'next/link';

interface IDashboardEventProps {
  event: IEventModal;
  loading: boolean;
}

const DashboardEvent: React.FC<IDashboardEventProps> = ({ event, loading }) => {
  return (
    <Card style={{ marginTop: '12px' }}>
      <ProgressCardContent loading={loading}>
        <Link
          href="/dashboard/events/[id]"
          as={`/dashboard/events/${event.id}`}
        >
          <a>
            <Typography gutterBottom variant="h5" component="h2">
              {event.name}
            </Typography>
          </a>
        </Link>
        <Typography variant="body2" color="textSecondary" component="p">
          {event.description}
        </Typography>
        <Typography variant="body2" color="textSecondary" component="p">
          {event.capacity}
        </Typography>
      </ProgressCardContent>
    </Card>
  );
};

export default DashboardEvent;
