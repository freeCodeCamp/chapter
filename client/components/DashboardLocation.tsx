import React, { useState } from 'react';
import {
  Card,
  Typography,
  Button,
  makeStyles,
  Link as MULink,
} from '@material-ui/core';
import { useDispatch } from 'react-redux';
import Link from 'next/link';

import ProgressCardContent from './ProgressCardContent';
import { ILocationModal } from 'client/store/types/locations';
import { locationActions } from 'client/store/actions';

interface IDashboardLocationProps {
  location: ILocationModal;
  loading: boolean;
}

const useStyles = makeStyles(() => ({
  bottom: {
    display: 'flex',
    alignItems: 'center',
  },
  action: {
    marginLeft: '20px',
  },
}));

const DashboardLocation: React.FC<IDashboardLocationProps> = ({
  location,
  loading,
}) => {
  const [confirm, setConfirm] = useState<boolean>(false);
  const dispatch = useDispatch();

  const styles = useStyles();

  const remove = () => {
    dispatch(locationActions.remove(location.id));
  };

  return (
    <Card style={{ marginTop: '12px' }}>
      <ProgressCardContent loading={loading}>
        <MULink gutterBottom variant="h5" component="h2">
          <Link href={`/dashboard/locations/${location.id}`}>
            {location.city}
          </Link>
        </MULink>
        <Typography variant="body2" color="textSecondary" component="p">
          {`${location.region}, ${location.country_code}, ${location.postal_code}`}
        </Typography>
        <Typography variant="body2" color="textSecondary" component="p">
          {location.address}
        </Typography>

        <div className={styles.bottom}>
          {confirm ? (
            <Button onClick={remove}>Are you sure</Button>
          ) : (
            <Button onClick={() => setConfirm(true)}>DELETE</Button>
          )}

          <Link href={`/dashboard/locations/${location.id}/edit`}>
            <a className={styles.action}>Edit</a>
          </Link>
        </div>
      </ProgressCardContent>
    </Card>
  );
};

export default DashboardLocation;
