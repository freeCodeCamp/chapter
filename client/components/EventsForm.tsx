import React from 'react';
import { useForm, Controller } from 'react-hook-form';
import { FormControl, TextField, Button, makeStyles } from '@material-ui/core';

interface IField {
  key: string;
  label: string;
  placeholder?: string;
  type: string;
  defaultValue?: string;
}

const fields: IField[] = [
  {
    key: 'name',
    type: 'text',
    label: 'Event title',
    placeholder: 'Foo and the Bars',
  },
  {
    key: 'description',
    type: 'text',
    label: 'Description',
    placeholder: '',
  },
  {
    key: 'capacity',
    type: 'number',
    label: 'Capacity',
    placeholder: '50',
  },
  {
    key: 'tags',
    type: 'text',
    label: 'Tags (separated by a comma)',
    placeholder: 'Foo, bar',
  },
  {
    key: 'start_at',
    type: 'datetime-local',
    label: 'Start at',
    defaultValue: new Date().toISOString().slice(0, 16),
  },
  {
    key: 'end_at',
    type: 'datetime-local',
    label: 'End at',
    defaultValue: new Date(Date.now() + 1000 * 60 * 60)
      .toISOString()
      .slice(0, 16),
  },
];

const useStyles = makeStyles(() => ({
  form: {
    display: 'flex',
    flexDirection: 'column',
    maxWidth: '25%',
  },
  item: {
    marginTop: '20px',
  },
}));

export interface IEventFormData {
  name: string;
  description: string;
  capacity: number;
  tags: string;
}

interface IEventsFormProps {
  onSubmit: (data: IEventFormData) => void;
}

const EventsForm: React.FC<IEventsFormProps> = ({ onSubmit }) => {
  const { control, handleSubmit } = useForm();
  const styles = useStyles();

  return (
    <form onSubmit={handleSubmit(onSubmit)} className={styles.form}>
      {fields.map(field => (
        <FormControl className={styles.item} key={field.key}>
          <Controller
            control={control}
            as={
              <TextField
                label={field.label}
                type={field.type}
                placeholder={field.placeholder}
              />
            }
            name={field.key}
            defaultValue={field.defaultValue}
            options={{ required: true }}
          />
        </FormControl>
      ))}
      <Button
        className={styles.item}
        variant="contained"
        color="primary"
        type="submit"
      >
        Add Event
      </Button>
    </form>
  );
};

export default EventsForm;
