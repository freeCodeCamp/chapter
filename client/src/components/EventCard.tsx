import { LockIcon } from '@chakra-ui/icons';
import { Tag, Box, Flex, Image, Grid, GridItem, Text } from '@chakra-ui/react';
import { Link } from 'chakra-next-link';
import { isPast } from 'date-fns';
import React from 'react';
import { Chapter, Event } from '../generated/graphql';
import { formatDate } from '../util/date';

type EventCardProps = {
  event: Pick<
    Event,
    | 'id'
    | 'name'
    | 'description'
    | 'start_at'
    | 'ends_at'
    | 'image_url'
    | 'invite_only'
    | 'canceled'
  > & {
    chapter: Pick<Chapter, 'id' | 'name'>;
  };
};

export const EventCard: React.FC<EventCardProps> = ({ event }) => {
  const metaTag = (
    <>
      {event.invite_only && (
        <Tag
          borderRadius="lg"
          mt="1"
          paddingInline="[1 , 2]"
          paddingBlock="[.5, 1]"
          colorScheme={'blue'}
          fontSize={['small', 'md']}
          maxWidth={'8em'}
          maxH={'2em'}
        >
          <LockIcon marginRight=".25rem" />
          Invite only
        </Tag>
      )}
    </>
  );
  enum EventStatus {
    canceled = 'Canceled',
    running = 'Running',
    ended = 'Ended',
    upcoming = 'Upcoming',
  }

  const statusToStyle = {
    [EventStatus.canceled]: { 'data-cy': 'event-canceled', color: 'red.500' },
    [EventStatus.running]: {
      color: 'gray.00',
      backgroundColor: 'gray.45',
      paddingInline: '.3em',
      borderRadius: 'sm',
    },
    [EventStatus.ended]: { color: 'gray.45', fontWeight: '400' },
    [EventStatus.upcoming]: {},
  };

  const hasEnded = isPast(new Date(event.ends_at));
  const getEventStatus = ({
    canceled,
    hasStarted,
    hasEnded,
  }: {
    canceled: boolean;
    hasStarted: boolean;
    hasEnded: boolean;
  }) => {
    if (canceled) return EventStatus.canceled;
    if (hasEnded) return EventStatus.ended;
    if (hasStarted) return EventStatus.running;
    return EventStatus.upcoming;
  };
  const eventStatus = getEventStatus({
    canceled: event.canceled,
    hasStarted: isPast(new Date(event.start_at)),
    hasEnded,
  });
  const eventStatusStyle = statusToStyle[eventStatus];
  return (
    <Flex
      borderWidth="1px"
      borderRadius="lg"
      overflow="hidden"
      width={'full'}
      {...(hasEnded && { opacity: 0.6 })}
      position="relative"
    >
      <Box w="240px" h="120px" display={['none', 'block']} background="gray.85">
        <Image
          width="100%"
          height="100%"
          src={
            event.image_url ||
            'https://cdn.freecodecamp.org/chapter/brown-curtain-small.jpg'
          }
          fit="cover"
          fallbackSrc="https://cdn.freecodecamp.org/chapter/brown-curtain-small.jpg"
          fallbackStrategy="onError"
        />
      </Box>
      <Box p="3" py={3} width="full" data-cy="event-card">
        <Grid
          mb="2"
          gridTemplateColumns={'repeat(3, 1fr)'}
          templateAreas={`
          "eventname eventname eventname"
          "chaptername chaptername chaptername"
          "eventstart eventstart eventstart"
          "metatag metatag metatag"
          `}
        >
          <Link
            data-cy="event-link"
            mt={1}
            size="sm"
            gridArea={'eventname'}
            fontSize={'xl'}
            fontWeight={700}
            href={`/events/${event.id}`}
            _before={{
              content: '""',
              position: 'absolute',
              inset: '0',
              zIndex: '1',
              width: '100%',
              height: '100%',
              cursor: 'pointer',
              textDecoration: 'underline',
            }}
          >
            {event.name}
          </Link>
          <GridItem area={'metatag'}>{metaTag}</GridItem>
          <Text
            opacity={'.8'}
            gridArea={'eventstart'}
            marginBottom={['1', '2']}
            {...eventStatusStyle}
            fontSize={['smaller', 'sm']}
            fontWeight={'semibold'}
          >
            {eventStatus}: {formatDate(event.start_at)}
          </Text>
        </Grid>
      </Box>
    </Flex>
  );
};
