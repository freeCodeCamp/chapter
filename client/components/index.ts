import SomeComponent from './SomeComponent';
import AddSponsor from './AddSponsor';
import ProgressCardContent from './ProgressCardContent';

export { SomeComponent, AddSponsor, ProgressCardContent };
