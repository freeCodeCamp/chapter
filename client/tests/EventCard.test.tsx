/**
 * @jest-environment jsdom
 */

import { render } from '@testing-library/react';
import React from 'react';
import { EventCard } from '../src/components/EventCard';

describe('EventCard', () => {
  it('should render', () => {
    const { container } = render(
      <EventCard
        event={{
          chapter: { id: 1, name: 'foo' },
          id: 1,
          name: 'bar',
          description: 'baz',
          start_at: 123,
          ends_at: 123,
          invite_only: true,
          canceled: false,
          image_url: 'http://example.com/image.png',
        }}
      />,
    );
    expect(container).toMatchSnapshot();
  });
});
