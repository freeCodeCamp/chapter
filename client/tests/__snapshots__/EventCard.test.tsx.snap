// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventCard should render 1`] = `
<div>
  <div
    class="css-dui41h"
  >
    <div
      class="css-1r4lq52"
    >
      <img
        class="chakra-image css-bh4wo8"
        src="http://example.com/image.png"
      />
    </div>
    <div
      class="css-1fuobds"
      data-cy="event-card"
    >
      <div
        class="css-pfvov2"
      >
        <a
          class="chakra-link css-1erir9l"
          data-cy="event-link"
          href="/events/1"
        >
          bar
        </a>
        <div
          class="css-14ygirk"
        >
          <span
            class="css-y2lvzo"
          >
            <svg
              class="chakra-icon css-1txi66j"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19.5,9.5h-.75V6.75a6.75,6.75,0,0,0-13.5,0V9.5H4.5a2,2,0,0,0-2,2V22a2,2,0,0,0,2,2h15a2,2,0,0,0,2-2V11.5A2,2,0,0,0,19.5,9.5Zm-9.5,6a2,2,0,1,1,3,1.723V19.5a1,1,0,0,1-2,0V17.223A1.994,1.994,0,0,1,10,15.5ZM7.75,6.75a4.25,4.25,0,0,1,8.5,0V9a.5.5,0,0,1-.5.5H8.25a.5.5,0,0,1-.5-.5Z"
                fill="currentColor"
              />
            </svg>
            Invite only
          </span>
        </div>
        <p
          class="chakra-text css-1q8l8r5"
        >
          Ended
          : 
          Thu, Jan 1 @ 00:00
        </p>
      </div>
    </div>
  </div>
</div>
`;
